ytypes/node_cache.go

```diff
@@ -0,0 +1,274 @@
+		return
+	}
+
+	err = status.Errorf(
+		codes.Unknown,
+		"failed to retrieve node, value %v",
+		val,
+	)
+
+	return
+}
+
+// update performs `NodeCache` update based on the input arguments.
+func (c *NodeCache) update(nodes []*TreeNode, tp, np *gpb.Path, parent, root interface{}, pathStr *string) {
+	var pathRep string
+	if pathStr != nil {
+		pathRep = *pathStr
+	} else {
+		if tp != nil && len(tp.GetElem()) > 0 {
+			var err error
+
+			pathRep, err = uniquePathRepresentation(appendElem(np, tp.GetElem()[0]))
+			if err != nil {
+				return
+			}
+		} else {
+			var err error
+
+			pathRep, err = uniquePathRepresentation(np)
+			if err != nil {
+				return
+			}
+		}
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if _, ok := c.store[pathRep]; !ok {
+		c.store[pathRep] = &cachedNodeInfo{}
+	}
+
+	if c.store[pathRep].nodes == nil || len(c.store[pathRep].nodes) == 0 {
+		c.store[pathRep].nodes = nodes
+	} else {
+		// Only update the data.
+		c.store[pathRep].nodes[0].Data = nodes[0].Data
+	}
+
+	c.store[pathRep].parent = parent
+	c.store[pathRep].root = root
+}
+
+// delete removes the path entry from the node cache.
+func (c *NodeCache) delete(path *gpb.Path) {
+	// Delete in the nodeCache.
+	nodePath, err := uniquePathRepresentation(path)
+	if err != nil {
+		return
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	keysToDelete := []string{}
+	for k := range c.store {
+		if strings.Contains(k, nodePath) {
+			keysToDelete = append(keysToDelete, k)
+		}
+	}
+
+	for _, k := range keysToDelete {
+		delete(c.store, k)
+	}
+}
+
+// get tries to retrieve the cached `TreeNode` slice. If the cache doesn't contain the
+// target `TreeNode` slice or the cached data is `nil`, an error is returned.
+func (c *NodeCache) get(path *gpb.Path) ([]*TreeNode, error) {
+	pathRep, err := uniquePathRepresentation(path)
+	if err != nil {
+		return nil, err
+	}
+
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	if nodeInfo, ok := c.store[pathRep]; ok {
+		ret := nodeInfo.nodes
+		if ret == nil || len(ret) == 0 {
+			return nil, status.Error(codes.NotFound, "cache: no node was found.")
+		} else if util.IsValueNil(ret[0].Data) {
+			return nil, status.Error(codes.NotFound, "cache: nil value node was found.")
+		}
+
+		return ret, nil
+	}
+
+	return nil, nil
+}
+
+// uniquePathRepresentation returns the unique path representation. The current
+// implementation uses json marshal for both simplicity and performance.
+//
+// https://pkg.go.dev/encoding/json#Marshal
+//
+// Map values encode as JSON objects. The map's key type must either be a string,
+// an integer type, or implement encoding.TextMarshaler. The map keys are sorted
+// and used as JSON object keys by applying the following rules, subject to the
+// UTF-8 coercion described for string values above:
+//
+//   - keys of any string type are used directly
+//
+//   - encoding.TextMarshalers are marshaled
+//
+//   - integer keys are converted to strings
+func uniquePathRepresentation(path *gpb.Path) (string, error) {
```

Comment:
Might need to compare the performance of these two if that matters. Previously `path.String` (`encoding/prototext/encode.go`) was used. Its performance was more than doubled after switching to `json.Marshal` (`json.Marshal` is deterministic so it's fine to use). After switching to `go-json` its was again largely improved.

Because computing the unique path representation is one of the most expensive, if not the most expensive parts of the node cache, the performance of this function affects the performance of the node cache.